### PR TITLE
OCL: Data race fix between oclCleanupCallback and Mat::GetUMat();

### PR DESCRIPTION
**Pull request by lupustr3**
Case 1: reuse of old src Mat in OCL pipe leads to data race with oclCleanupCallback and can result in asser fail or other undefined behavior.

Case 2: concurent deallocation of UMatData between UMat (oclCleanupCallback) and Mat (stack) deallocators.

EventCnt object was added to UMatData to track status of oclCleanupCallback;
Two test cases were implemented to reproduce the problem;

### CODE REVIEW

**Comment by lupustr3** [comment]
I have following message from build system:
 "Failed
  No regression data for src argument"

Can you help me to add regression data for my test?

File: modules/core/perf/opencl/perf_matop.cpp [hunk 1]
```diff
@@ -13,6 +13,59 @@
...
+    declare.in(src, WARMUP_RNG).out(dst);
+
+    OCL_TEST_CYCLE_MULTIRUN(1000)
+    {
+        UMat tmpUMat = src.getUMat(ACCESS_RW);
+        tmpUMat.convertTo(dst, dtype);
+        ::cv::ocl::finish();
+    }
+
+    SANITY_CHECK(src);
```
Comment (alalek) [inline modules/core/perf/opencl/perf_matop.cpp]: This check requires test data about src values in `opencv_extra` repository. This check looks useless, so use SANITY_CHECK_NOTHING here.

Is there stable result (I mean time measurement for this performance test)? If no, please move tests to accuracy.

Comment (lupustr3) [inline modules/core/perf/opencl/perf_matop.cpp]: > This check requires test data about src values in opencv_extra repository. This check looks useless, so use SANITY_CHECK_NOTHING here.

Due to this data race there is scenario where src data become corrupted but no other fails happens, so I use this as additional check.

> Is there stable result (I mean time measurement for this performance test)? If no, please move tests to accuracy.

Ok.

**Comment by vpisarev** [comment]
@lupustr3, thanks for the patch! With this patch the overhead for each UMat becomes noticeably bigger (separate sync primitive for each UMat etc.). I'm afraid, it does not work for us. If the problem is really serious and common, I would like to see some lighter, more efficient solution for it.

**Comment by vpisarev** [comment]
ok, let's close it for now

**Pull request closed without merging**